File: games/maiya/tools/regression.py
"""Headless movement, climbing, arena and bonus checks using ELF debug symbols."""

import argparse
import json
import os
from pathlib import Path
import subprocess

from build import GAME, ROOT, WORK


def main():
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument("--seconds", type=int, default=50)
    parser.add_argument("--mame", default="mame")
    parser.add_argument("--scenario", choices=("idle", "walk", "climb", "boss", "bonus", "continue",
                                              "continue-exit", "continue-timeout", "tray", "factory",
                                              "pickups", "pit"), default="walk")
    parser.add_argument("--idle", action="store_true", help="Capture startup without gameplay inputs")
    parser.add_argument("--output", type=Path)
    parser.add_argument("--eagle-bios", action="store_true")
    parser.add_argument("--platform", choices=("mvs", "aes"), default="mvs")
    args = parser.parse_args()
    marker = WORK / "platform.txt"
    built_platform = marker.read_text().strip() if marker.exists() else "mvs"
    if built_platform != args.platform:
        parser.error(f"Build with --platform {args.platform} first")
    if args.scenario == "continue-exit" and args.platform != "aes":
        parser.error("The Exit choice is only offered on AES")
    if args.idle:
        args.scenario = "idle"
    output = (args.output or GAME / "build" / args.scenario).resolve()
    output.mkdir(parents=True, exist_ok=True)
    gdb = next(ROOT.parent.glob("x-tools-v*/m68k-unknown-elf/bin/m68k-unknown-elf-gdb"))
    fields = {name: "&mg." + name for name in (
        "stage", "state", "player", "boss", "boss_active", "state_timer",
        "gate_unlocked", "climbing", "demo", "shake_x", "lives", "continues",
        "has_key", "next_stage", "art", "coins", "flowers", "critters",
        "swift", "might", "veil", "spring", "crown", "items", "pick_mask", "hud_dirty")}
    fields["camera_x"] = "&mg.camera.x"
    # BIOS_PLAYER1_MODE (0x10FDB6), the same address regression.lua already
    # reads as "player_mode": Maiya no longer keeps a Start-press counter of
    # her own, since the BIOS's own mode flag is the live, race-free signal.
    fields["start_count"] = "0x10FDB6"
    fields.update({"char_" + name: f"&((NGCharacter*)0)->{name}"
                   for name in ("x", "y", "x_fp", "y_fp", "hp", "flip_x", "data0")})
    fields.update({"item_" + name: f"&((MGItem*)0)->{name}" for name in ("life", "key", "source")})
    fields["item_size"] = "sizeof(MGItem)"
    command = [str(gdb), "-batch", str(WORK / "out/game")]
    for name, expr in fields.items():
        command += ["-ex", f'printf "{name}=%lu\\n", (unsigned long){expr}']
    layout = dict(line.split("=", 1) for line in subprocess.check_output(command, text=True).splitlines() if "=" in line)
    (output / "layout.lua").write_text("return {" + ",".join(f"{key}={int(value)}" for key, value in layout.items()) + "}\n")
    env = dict(os.environ, SDL_VIDEODRIVER="dummy", SDL_AUDIODRIVER="dummy", DISPLAY="",
               MG_CAPTURE_DIR=str(output), MG_SCENARIO=args.scenario, MG_PLATFORM=args.platform)
    bios_path = str(ROOT / "bios/test_roms") + ";" if args.eagle_bios else ""
    command = [args.mame, "aes" if args.platform == "aes" else "neogeo", "-noreadconfig", "-rompath",
               f"{bios_path}{WORK / 'roms'};{ROOT / 'roms'}", "-hashpath",
               str(WORK / "hash_eagle/maiya"), "-cart1", "maiya", "-bios", "asia" if args.platform == "aes" else "euro",
               "-video", "none", "-sound", "none", "-nothrottle", "-nonvram_save",
               "-seconds_to_run", str(args.seconds), "-skip_gameinfo",
               "-cfg_directory", str(output / "cfg"), "-nvram_directory", str(output / "nvram"),
               "-snapshot_directory", str(output), "-autoboot_delay", "0",
               "-autoboot_script", str(GAME / "tools/regression.lua")]
    with (output / "mame.log").open("w") as log:
        subprocess.run(command, env=env, cwd=output, stdout=log, stderr=subprocess.STDOUT, check=True, timeout=300)
    samples = [json.loads(line) for line in (output / "telemetry.jsonl").read_text().splitlines()]
    active = [s for s in samples if s["mode"] == 2 and s["player"] and not s["demo"]]
    if args.scenario == "idle":
        active = [s for s in samples if s["player"]]
    assert active, "The cartridge never entered play"
    if args.scenario in ("walk", "climb", "boss", "bonus", "continue"):
        assert max(s["x"] for s in active) > min(s["x"] for s in active), "Player did not move"
    if args.scenario == "climb":
        climbing = [s for s in active if s["climbing"]]
        assert climbing and max(s["y"] for s in climbing) - min(s["y"] for s in climbing) > 50
        assert len({s["flip"] for s in climbing}) == 1, "Climbing unexpectedly reverses facing"
    elif args.scenario == "boss":
        arena = [s for s in active if s["boss_active"] and s["boss"]]
        assert arena, "Guardian never spawned"
        assert max(s["boss_x"] for s in arena) - min(s["boss_x"] for s in arena) > 40, "Guardian does not walk"
        assert all(3540 <= s["x"] <= 3820 for s in arena), "Player escaped arena"
        settled = arena[4:]
        assert settled and all(s["camera_x"] - s["shake_x"] == 3520 for s in settled), "Arena camera did not lock"
    elif args.scenario == "bonus":
        bonus = [s for s in active if s["state"] == 3]
        assert bonus and all(s["camera_x"] == 0 for s in bonus), "Bonus reused the previous arena"
        assert any(s["stage"] == 2 and s["state"] == 1 for s in active), "Bonus did not advance to mission 3"
    elif args.scenario == "continue":
        waiting = [s for s in active if s["state"] == 5 and s["continues"] == 3]
        assert waiting and all(s["player_mode"] == 2 for s in waiting), "BIOS was not told about Continue"
        if args.platform == "mvs":
            assert any(s["credit"] == 1 for s in waiting), "Coin did not reach Continue"
        resumed = [s for s in active if s["continues"] == 2 and s["state"] in (0, 1)]
        assert resumed, "Start did not resume the game"
        first = resumed[0]
        # A continue restarts the whole mission, not the spot she fell: the
        # same head-of-road entrance every fresh arrival uses, and whatever
        # she was carrying goes with the reset, the same as any other retry.
        assert first["x"] == 64 and first["y"] < 192, "Respawn did not return to the start of the mission"
        assert first["lives"] == 3, "Continue must restore three lives"
        if args.platform == "mvs":
            assert first["credit"] == 0, "Continue must spend exactly one credit"
        assert not first["has_key"] and not first["gate_unlocked"], "Continue should reset mission progress"
        assert any(s["state"] == 1 for s in resumed), "Respawn never returned control after landing"
    elif args.scenario in ("continue-exit", "continue-timeout"):
        waiting = [s for s in active if s["state"] == 5]
        assert waiting, "Continue was not offered"
        assert any(s["time"] > waiting[-1]["time"] and
                   (s["player_mode"] == 3 or s["demo"] or s["mode"] != 2)
                   for s in samples), "Exit did not end the player's run"
        assert not any(s["continues"] == 2 for s in active), "Exit spent a continuation"
    elif args.scenario == "factory":
        factory = [s for s in active if s["stage"] == 6 and s["state"] == 1]
        assert factory and any(s["boss_active"] for s in factory), "Factory guardian did not spawn"
        assert any(s["boss_kind"] == 5 for s in factory), "Second factory guardian did not spawn"
    elif args.scenario == "pickups":
        assert any(s["pick_mask"] & 2 for s in active), "Later road pickup never became collectible"
        assert any(s["pick_mask"] & 16 for s in active), "Uncollected shelf pickup was lost on backtracking"
        assert all(s["key_count"] == 1 for s in active if s["state"] == 1), "Mandatory key was discarded or duplicated"
        assert all(s["duplicate_items"] == 0 for s in active), "Map pickup spawned twice"
    elif args.scenario == "pit":
        falls = [s for s in active if s["stage"] == 1]
        assert falls, "Never reached Valley of Falls"
        assert max(s["x"] for s in falls) > 590, "Never walked past the first pit"
        assert any(s["state"] == 4 for s in falls), "Falling into the pit never registered a death"
        assert any(s["lives"] == 2 for s in falls), "Falling into the pit did not cost exactly one life"
    assert "LUA ERROR" not in (output / "mame.log").read_text(), "Capture script failed"
    print(f"PASS {args.scenario}: {output}")


if __name__ == "__main__":
    main()

File: games/maiya/tools/regression.lua
-- Inputs and debugger-only scenario placement; no test hooks in the cartridge.
local machine = manager.machine
local screen = machine.screens[':screen']
local memory = machine.devices[':maincpu'].spaces['program']
local output = assert(os.getenv('MG_CAPTURE_DIR'))
local a = dofile(output .. '/layout.lua')
local scenario = os.getenv('MG_SCENARIO') or 'walk'
local aes = os.getenv('MG_PLATFORM') == 'aes'
local testing_continue = scenario:sub(1, 8) == 'continue'
local log = assert(io.open(output .. '/telemetry.jsonl', 'w'))
local next_capture, placed, play_frame, continue_frame = 0, false, 0, 0
local pit_step, pit_wait = 0, 0
local pit_after, pit_after_frame = false, 0
local function u8(name) return memory:read_u8(a[name]) end
local function w8(name, value) memory:write_u8(a[name], value) end
local function s16(addr)
    local n = memory:read_u16(addr)
    return n >= 32768 and n - 65536 or n
end
local function input(name, value)
    for _, port in pairs(machine.ioport.ports) do
        local f = port.fields[name]
        if f then f:set_value(value and 1 or 0) end
    end
end
local function position(p, x, y)
    memory:write_u16(p + a.char_x, x)
    memory:write_u16(p + a.char_y, y)
    memory:write_u32(p + a.char_x_fp, x * 256)
    memory:write_u32(p + a.char_y_fp, y * 256)
end
local function start(value)
    input('1 Player Start', value)
    input('P1 Start', value)
end
emu.register_frame_done(function()
    local t = machine.time:as_double()
    input('Coin 1', t > 8 and t < 8.25)
    start(t > 12 and t < 12.25)
    local p = memory:read_u32(a.player)
    if p < 0x100000 or p > 0x10efff then p = 0 end
    -- Between Start landing and real gameplay there are now three screens
    -- (hero-select, how-to-play, the intro story) that each need a real
    -- button, not just a direction -- mg.player does not exist until she is
    -- through all three. Push through them here, unconditionally; once p is
    -- valid this stops, so a stray press from this schedule can never land
    -- during a scenario's own scripted inputs below.
    if p == 0 then
        input('P1 A', (t > 13.0 and t < 13.15) or (t > 14.5 and t < 14.65) or
                       (t > 16.0 and t < 16.15) or (t > 17.5 and t < 17.65))
    end
    local mode = memory:read_u8(0x10fdaf)
    if p ~= 0 and mode == 2 and u8('demo') == 0 then
        play_frame = play_frame + 1
        if not placed and u8('state') == 1 then
            placed = true
            if scenario == 'climb' then position(p, 708, 192) end
            if scenario == 'boss' then
                position(p, 3560, 192)
                memory:write_u16(a.camera_x, 3400)
                w8('gate_unlocked', 1)
            end
            if scenario == 'bonus' then
                w8('stage', 1)
                w8('state', 2)
                memory:write_u16(a.state_timer, 1)
            end
            if testing_continue then
                -- She always restarts a mission at its own head now, so
                -- there is no mid-level checkpoint to keep in sync here;
                -- only her position at the moment of death is staged.
                position(p, 1500, 192)
                w8('has_key', 1)
                w8('gate_unlocked', 1)
                w8('lives', 0)
                w8('state', 4)
                memory:write_u16(a.state_timer, 1)
            end
            if scenario == 'tray' then
                w8('has_key', 1)
                w8('coins', 99)
                w8('flowers', 99)
                w8('critters', 99)
                for _, name in ipairs({'swift', 'might', 'veil', 'spring', 'crown'}) do
                    memory:write_u16(a[name], 480)
                end
            end
            if scenario == 'factory' then
                w8('next_stage', 6)
                w8('state', 8)
                memory:write_u16(a.state_timer, 1)
            end
            -- Valley of Falls, same way the game reaches it after a boss:
            -- the ordinary mission-to-mission interlude, not a shortcut
            -- wired into the cartridge.
            if scenario == 'pit' then
                w8('next_stage', 1)
                w8('state', 8)
                memory:write_u16(a.state_timer, 1)
            end
        end
        -- Once the interlude has dropped her into Valley of Falls, drive a
        -- fixed sequence past its first pit (x=560, width=48): approach and
        -- stand at the lip, jump it, then a second pass that walks straight
        -- in without jumping, to confirm the fall (and the one-life cost)
        -- actually happens.
        if scenario == 'pit' and u8('stage') == 1 and u8('state') == 1 and p ~= 0 then
            local x = s16(p + a.char_x)
            if pit_step == 0 then
                position(p, 460, 192)
                memory:write_u8(p + a.char_hp, 5)
                pit_step, pit_wait = 1, 0
            elseif pit_step == 1 then
                input('P1 Right', true)
                if x >= 548 then pit_step, pit_wait = 15, 0 end
            elseif pit_step == 15 then
                input('P1 Right', false)   -- stop and stand at the lip first
                pit_wait = pit_wait + 1
                if pit_wait == 15 then screen:snapshot(output .. '/pit_01_at_lip.png') end
                if pit_wait == 20 then pit_step, pit_wait = 2, 0 end
            elseif pit_step == 2 then
                input('P1 Right', true)
                pit_wait = pit_wait + 1
                if pit_wait == 6 then input('P1 A', true) end
                if pit_wait == 12 then input('P1 A', false) end
                if pit_wait == 20 then screen:snapshot(output .. '/pit_02_mid_jump.png') end
                if pit_wait == 40 then
                    screen:snapshot(output .. '/pit_03_after_jump.png')
                    pit_step, pit_wait = 3, 0
                end
            elseif pit_step == 3 then
                input('P1 Right', false)
                position(p, 460, 192)
                memory:write_u8(p + a.char_hp, 5)
                w8('lives', 3)
                pit_step, pit_wait = 4, 0
            elseif pit_step == 4 then
                input('P1 Right', true)
                if x >= 555 then
                    screen:snapshot(output .. '/pit_04_before_fall.png')
                    pit_step, pit_wait = 5, 0
                end
            elseif pit_step == 5 then
                input('P1 Right', true)
                pit_wait = pit_wait + 1
                if pit_wait == 24 then screen:snapshot(output .. '/pit_05_falling.png') end
            end
        end
        -- The fall interrupts state==1, so the branch above stops running
        -- (and pit_step freezes at 5) the instant she dies; catch the
        -- after-death and respawn shots here instead, outside that gate.
        if scenario == 'pit' and pit_step == 5 and u8('state') == 4 and not pit_after then
            pit_after, pit_after_frame = true, play_frame
        end
        if pit_after and play_frame - pit_after_frame == 40 then
            screen:snapshot(output .. '/pit_06_after_fall.png')
        end
        if pit_after and play_frame - pit_after_frame == 200 then
            screen:snapshot(output .. '/pit_07_respawned.png')
        end
        input('P1 Right', scenario == 'walk' or (scenario == 'bonus' and play_frame % 240 < 120))
        input('P1 Left', scenario == 'boss' or (scenario == 'bonus' and play_frame % 240 >= 120))
        input('P1 Up', scenario == 'climb' and placed and play_frame < 210)
        input('P1 Down', scenario == 'climb' and play_frame >= 240 and play_frame < 390)
        input('P1 A', (scenario == 'walk' or scenario == 'bonus') and play_frame % 180 < 6)
        input('P1 B', (scenario == 'walk' or scenario == 'bonus') and play_frame % 24 < 6)
        input('P1 D', false)
        if scenario == 'pickups' and placed then
            local x, y = 901, 192
            if play_frame > 180 then x, y = 528, 104 end
            if play_frame > 300 then x, y = 901, 192 end
            position(p, x, y)
            memory:write_u8(p + a.char_hp, 5)
        end
        if testing_continue and u8('state') == 5 then
            continue_frame = continue_frame + 1
            if scenario == 'continue-timeout' then
                start(false)
            elseif aes then
                input('P1 Right', scenario == 'continue-exit' and continue_frame >= 45 and continue_frame < 55)
                input('P1 A', continue_frame >= 95 and continue_frame < 105)
            else
                input('Coin 1', continue_frame >= 45 and continue_frame < 55)
                input('P1 A', continue_frame >= 65 and continue_frame < 75)
                start((continue_frame >= 15 and continue_frame < 25) or
                      (continue_frame >= 95 and continue_frame < 105))
            end
        end
        if scenario == 'factory' and u8('stage') == 6 and u8('state') == 1 then
            position(p, 4840, 192)
            w8('gate_unlocked', 1)
            memory:write_u8(p + a.char_hp, 5)
            local boss = memory:read_u32(a.boss)
            if boss >= 0x100000 and boss <= 0x10efff and memory:read_u16(boss + a.char_data0) == 6 then
                memory:write_u8(boss + a.char_hp, 1)
                input('P1 D', play_frame % 90 < 10)
            end
        end
        -- Keep arena observation alive without changing movement or collisions.
        if scenario == 'boss' then memory:write_u8(p + a.char_hp, 5) end
    end
    if t >= next_capture then
        local b = memory:read_u32(a.boss)
        if b < 0x100000 or b > 0x10efff then b = 0 end
        local key_count, duplicates, sources = 0, 0, {}
        for i = 0, 3 do
            local item = a.items + i * a.item_size
            if memory:read_u8(item + a.item_life) ~= 0 then
                key_count = key_count + memory:read_u8(item + a.item_key)
                local source = memory:read_u8(item + a.item_source)
                if source ~= 0 then
                    if sources[source] then duplicates = duplicates + 1 end
                    sources[source] = true
                end
            end
        end
        screen:snapshot(string.format('%s/frame_%05d.png', output, math.floor(t * 10)))
        log:write(string.format('{"time":%.2f,"mode":%d,"state":%d,"stage":%d,"player":%d,"x":%d,"y":%d,"flip":%d,"climbing":%d,"camera_x":%d,"shake_x":%d,"boss_active":%d,"boss":%d,"boss_x":%d,"lives":%d,"continues":%d,"credit":%d,"player_mode":%d,"start_count":%d,"has_key":%d,"gate_unlocked":%d,"demo":%d,"pick_mask":%d,"key_count":%d,"duplicate_items":%d,"boss_kind":%d}\n',
            t, mode, u8('state'), u8('stage'), p, p ~= 0 and s16(p + a.char_x) or 0,
            p ~= 0 and s16(p + a.char_y) or 0, p ~= 0 and memory:read_u8(p + a.char_flip_x) or 0,
            u8('climbing'), s16(a.camera_x), s16(a.shake_x), u8('boss_active'), b, b ~= 0 and s16(b + a.char_x) or 0,
            u8('lives'), u8('continues'), aes and 0 or memory:read_u8(0xd00034), memory:read_u8(0x10fdb6),
            u8('start_count'), u8('has_key'), u8('gate_unlocked'), u8('demo'), memory:read_u16(a.pick_mask),
            key_count, duplicates, b ~= 0 and memory:read_u16(b + a.char_data0) or -1))
        log:flush()
        next_capture = t + 0.5
    end
end)
emu.register_stop(function() log:close() end)
